src/common/TermNames.h

```diff
@@ -38,49 +43,138 @@ class TermNames {
 
-    auto begin() const noexcept { return names.begin(); }
-    auto end() const noexcept { return names.end(); }
+    // std::optional does not work with references so we must use pointers
+    std::vector<TermName> const * tryGetNamesForTerm(PTRef term) const {
+        if (auto it = termToNames.find(term); it != termToNames.end()) { return &it->second; }
+
+        return nullptr;
+    }
 
-    bool empty() const noexcept { return size() == 0; }
-    std::size_t size() const noexcept { return names.size(); }
+    TermName const * tryGetNameForTerm(PTRef term) const {
+        if (auto vecPtr = tryGetNamesForTerm(term)) { return &pickName(*vecPtr); }
+
+        return nullptr;
+    }
+
+    // Iterates over const names and their corresponding terms (according to value_type)
+    inline auto begin() const noexcept;
+    inline auto end() const noexcept;
+
+    bool empty() const noexcept { return scopedNames.empty(); }
+    std::size_t size() const noexcept { return scopedNames.size(); }
+
+    // A const view to just the names
+    inline auto names() const;
+    // A const view to just the terms
+    auto const & terms() const { return _terms; }
 
 protected:
     friend class MainSolver;
 
     using NameToTermMap = std::unordered_map<TermName, PTRef>;
     using TermToNamesMap = std::unordered_map<PTRef, std::vector<TermName>, PTRefHash>;
 
+    class Iterator {
+    public:
+        using NameIterator = ScopedVector<TermName>::const_iterator;
+        using TermIterator = std::vector<PTRef>::const_iterator;
+
+        explicit Iterator(NameIterator nit, TermIterator tit) : nameIterator{nit}, termIterator{tit} {}
+
+        value_type operator*() const { return {*nameIterator, *termIterator}; }
+
+        Iterator & operator++() {
+            ++nameIterator;
+            ++termIterator;
+            return *this;
+        }
+        Iterator operator++(int) {
+            auto it = *this;
+            operator++();
+            return it;
+        }
+
+        bool operator==(Iterator const & rhs) const noexcept { return nameIterator == rhs.nameIterator; }
+
+    protected:
+        NameIterator nameIterator;
+        TermIterator termIterator;
+    };
+
+    class NamesView;
+
     void pushScope() {
         if (isGlobal()) { return; }
-        names.pushScope();
+        scopedNames.pushScope();
     }
 
     void popScope() {
         if (isGlobal()) { return; }
-        names.popScope([this](TermName const & name) {
+        assert(scopedNames.size() == _terms.size());
+        scopedNames.popScope([this](TermName const & name) {
             auto it = nameToTerm.find(name);
             if (it == nameToTerm.end()) { return; }
             PTRef term = it->second;
-            assert(termToNames.find(term) != termToNames.end());
-            auto & names = termToNames.at(term);
-            names.erase(std::find(names.begin(), names.end(), name));
+            auto & names_ = const_cast<std::vector<TermName> &>(namesForTerm(term));
```

Comment:
Let's not use `const_cast` please.
We can have a (private?) method to return non-const ref to the names, no?

Reply:
This is a perfectly fine use-case of `const_cast` proposed by Scott Meyer:
https://stackoverflow.com/questions/856542/elegant-solution-to-duplicate-const-and-non-const-getters
- it is ok to cast away const within non-const member function

Private method would cause the following name clash:
1) In most cases, objects are treated as non-const - `namesForTerm` is const method but the object itself is often not const
2) Hence the overload resolution would choose the non-const variant of `namesForTerm` which is however private. It would require the user to cast the object to const in order to call the public version.

Hence we can introduce the function as private but with a different name (e.g. `_namesForTerm`) to avoid the undesired overload resolution.